components/sceneutil/shadow.cpp

```diff
@@ -0,0 +1,166 @@
+
+        if (Settings::Manager::getBool("allow shadow map overlap", "Shadows"))
+            mShadowSettings->setMultipleShadowMapHint(osgShadow::ShadowSettings::CASCADED);
+        else
+            mShadowSettings->setMultipleShadowMapHint(osgShadow::ShadowSettings::PARALLEL_SPLIT);
+
+        if (Settings::Manager::getBool("enable debug hud", "Shadows"))
+            mShadowTechnique->enableDebugHUD();
+        else
+            mShadowTechnique->disableDebugHUD();
+    }
+
+    void ShadowManager::disableShadowsForStateSet(osg::ref_ptr<osg::StateSet> stateset)
+    {
+        int numberOfShadowMapsPerLight = Settings::Manager::getInt("number of shadow maps", "Shadows");
+        int baseShadowTextureUnit = 8 - numberOfShadowMapsPerLight;
+        
+        osg::ref_ptr<osg::Image> fakeShadowMapImage = new osg::Image();
+        fakeShadowMapImage->allocateImage(1, 1, 1, GL_DEPTH_COMPONENT, GL_FLOAT);
+        *(float*)fakeShadowMapImage->data() = std::numeric_limits<float>::infinity();
+        osg::ref_ptr<osg::Texture> fakeShadowMapTexture = new osg::Texture2D(fakeShadowMapImage);
+        fakeShadowMapTexture->setShadowComparison(true);
+        fakeShadowMapTexture->setShadowCompareFunc(osg::Texture::ShadowCompareFunc::ALWAYS);
+        for (int i = baseShadowTextureUnit; i < baseShadowTextureUnit + numberOfShadowMapsPerLight; ++i)
+        {
+            stateset->setTextureAttributeAndModes(i, fakeShadowMapTexture, osg::StateAttribute::ON | osg::StateAttribute::OVERRIDE | osg::StateAttribute::PROTECTED);
+            stateset->addUniform(new osg::Uniform(("shadowTexture" + std::to_string(i - baseShadowTextureUnit)).c_str(), i));
+            stateset->addUniform(new osg::Uniform(("shadowTextureUnit" + std::to_string(i - baseShadowTextureUnit)).c_str(), i));
+        }
+    }
+
+    ShadowManager::ShadowManager(osg::ref_ptr<osg::Group> sceneRoot, osg::ref_ptr<osg::Group> rootNode, unsigned int outdoorShadowCastingMask, unsigned int indoorShadowCastingMask, Shader::ShaderManager &shaderManager) : mShadowedScene(new osgShadow::ShadowedScene),
+        mShadowTechnique(new MWShadowTechnique),
+        mOutdoorShadowCastingMask(outdoorShadowCastingMask),
+        mIndoorShadowCastingMask(indoorShadowCastingMask)
+    {
+        mShadowedScene->setShadowTechnique(mShadowTechnique);
+
+        mShadowedScene->addChild(sceneRoot);
+        rootNode->addChild(mShadowedScene);
+
+        mShadowSettings = mShadowedScene->getShadowSettings();
+        setupShadowSettings();
+
+        mShadowTechnique->setupCastingShader(shaderManager);
+
+        enableOutdoorMode();
+    }
+
+    Shader::ShaderManager::DefineMap ShadowManager::getShadowDefines()
+    {
+        if (!mEnableShadows)
+            return getShadowsDisabledDefines();
+
+        Shader::ShaderManager::DefineMap definesWithShadows;
+        definesWithShadows.insert(std::make_pair(std::string("shadows_enabled"), std::string("1")));
```

Comment:
I saw the light of emplace and saw that it was good.
```suggestion
        definesWithShadows.emplace(std::string("shadows_enabled"), std::string("1"));
```

And line 138
```cpp
        definesWithoutShadows.emplace(std::string("shadows_enabled"), std::string("0"));
```

Reply:
I think everything should probably be consistent, so depending on what we deem to be our preferred style, we should choose either:

* Everything is emplaced. Everything is `if (setting) ….emplace(…, "1"); else ….emplace(…, "0");`
* Everything is emplaced. Everything is `….emplace(…, setting ? "1" : "0");`
* Everything is `operator[]`ed. Everything is `if (setting) …[…] = "1"; else …[…] "0";`
* Everything is `operator[]`ed. Everything is `…[…] = setting ? "1" : "0";`

@Capostrophic @akortunov @psi29a What do we think?

Reply:
Personally I prefer a last option.

Reply:
How do we do it elsewhere in the code base?

Reply:
Lots of the code predates emplace being an option, so we're likely to have fallen back on `operator[]` as that's much more readable than insert. That doesn't necessarily mean it's better, though. Emplace is faster, but as this code only runs once, performance isn't a huge concern versus readability.

Reply:
Looks like emplace is 'zero-copy' while typical 'operator[]' assignment does a copy. This is a once-time thing however, so I don't see how that would have a impact on performance.

So now it is just a matter of style.

Reply:
Shall we just say for non-critical areas of code, `operator[]` assignment is fine (more readable).
For performance critical areas of code however: `ensure` is preferred. 

Is this good?

Reply:
They don't have the same functionality, though, which should be kept in mind. `emplace` (and `insert`) is no-op if there is a value with the same key in the map while `operator[]` is not and will update the value.
I don't have a preference, although ternary is better than no ternary.

Reply:
This region has been updated. Is everyone happy now?

Reply:
I'm pressing the resolve button now.